readthedocs/api/v3/views.py

```diff
@@ -384,30 +384,117 @@ def create(self, request, **kwargs):  # pylint: disable=arguments-differ
         return Response(data=data, status=code)
 
 
-class NotificationsViewSet(
+class NotificationsForUserViewSet(
+    APIv3Settings,
+    FlexFieldsMixin,
+    ListModelMixin,
+    RetrieveModelMixin,
+    UpdateMixin,
+    UpdateModelMixin,
+    GenericViewSet,
+):
+
+    """
+    Endpoint to return all the notifications related to the logged in user.
+
+    Hitting this endpoint while logged in will return notifications attached to:
+
+     - User making the request
+     - Organizations where the user is owner/member
+     - Projects where the user is admin/member
+    """
+
+    model = Notification
+    serializer_class = NotificationSerializer
+    queryset = Notification.objects.all()
+    permission_classes = (IsAuthenticated,)
+    filterset_class = NotificationFilter
+
+    # http://chibisov.github.io/drf-extensions/docs/#usage-with-generic-relations
+    def get_queryset(self):
+        user = Notification.objects.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(User),
+            attached_to_id=self.request.user.pk,
+        )
+
+        # NOTE: using ``Subquery`` work better regarding performance since it makes only one query to the databse,
+        # instead of doing one to get the ids and another one to get the notifications.
+        # It also works better for users that have many projects/organizations since we don't need to load the ids
+        # in memory and then dump them into the SQL query.
+        # Besides, it performs better when applying the filters via ``NotificationFilter``
+        project = Notification.objects.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(Project),
+            attached_to_id__in=Subquery(
+                AdminPermission.projects(
+                    self.request.user, admin=True, member=False
+                ).values("id")
+            ),
+        )
+        organization = Notification.objects.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(Organization),
+            attached_to_id__in=Subquery(
+                AdminPermission.organizations(
+                    self.request.user, admin=True, member=False
+                ).values("id")
+            ),
+        )
+
+        # Return all the notifications related to this user attached to:
+        # User, Project and Organization models where the user is admin.
+        return user | project | organization
```

Comment:
We will be surfacing this queryset inside templates as well, when we go to render the notifications available to the user. This might make more sense as a queryset/manager method, so we can use the same logic in both.

Reply:
Done in cf241b958